Improve cover z-index solution

renatho opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This is an improved solution for this original PR: https://github.com/WordPress/gutenberg/pull/66093

As part of this PR, I also thought of removing [this effect that adds the `has-modal-open` in the editor](https://github.com/WordPress/gutenberg/pull/66093/files#diff-884dae78f1676d454c049107e9b05b2403eb042457f77f836bd170d104396b8bR89-R97) because that wouldn't be needed for this purpose anymore. But thinking more, I kept it to keep the consistency there between the frontend and the editor.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To improve the solution and eliminate the dependencies between Cover and Navigation blocks (the class that Navigation adds being used in the Cover styles).

More context starting from [this comment](https://github.com/WordPress/gutenberg/pull/66093/files#r1804730040) in the thread.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

It deprecates the current version of the Cover block, changing the order of the elements, and removing the `z-index` of them.

It continues using the `has-modal-open` class for backward compatibility (solution introduced in https://github.com/WordPress/gutenberg/pull/66093), but when the code is updated it's not used anymore (after the save).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Before switching to this branch, add a Cover block to a post.
2. Inside the cover block, add a Navigation block.
3. After the previously created Cover block, add another Cover block.
4. Switch to this branch.
5. Visit the site in the frontend on small screens and also visit the editor. Open the menu, and make sure it's displayed over the whole content.
6. Edit something in the post, and save to update the deprecated version of the block.
7. Check that the frontend continues working properly.

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/98bdfda3-4608-4869-90b1-887b09eaab89

Since I continued seeing the same result in my tests, I reused the same video from https://github.com/WordPress/gutenberg/pull/66093 just to illustrate the tests.

renatho committed:
Remove not used variable

renatho committed:
Change the order of the elements to the expected layers order

So we don't need to set z-index in all the elements, creating the stacking context for the inner blocks, for example.

renatho committed:
Update z-index styles to be applied only for backward compatibility

renatho commented on packages/base-styles/_z-index.scss:
```diff
@@ -31,12 +31,14 @@ $z-layers: (
+	// These z-index are now used only for a deprecated version of the cover block.
+	".wp-block-cover__inner-container": 1, // InnerBlocks area inside cover image block.
 	".wp-block-cover.has-background-dim::before": 1, // Overlay area inside block cover need to be higher than the video background.
-	".block-library-cover__padding-visualizer": 2, // BoxControl visualizer needs to be +1 higher than .wp-block-cover.has-background-dim::before
```
I'm removing it again because the change discussed [here](https://github.com/WordPress/gutenberg/pull/65626#discussion_r1775333362) was [reverted](https://github.com/WordPress/gutenberg/pull/66074).

renatho commented on packages/block-library/src/cover/save.js:
```diff
@@ -88,7 +88,8 @@ export default function save( { attributes } ) {
 		},
-		getPositionClassName( contentPosition )
+		getPositionClassName( contentPosition ),
+		'has-no-zindex-in-inner-blocks'
```
I didn't like this name, but I couldn't find a better one. I though about a `has-natural-elements-order` or something like this, but the `has-no-zindex-in-inner-blocks` seems more self-explanatory about the purpose of this.

aaronrobertshaw replied:
Could we avoid the introduction of the CSS class entirely using sibling CSS selectors?

This would also mean that only blocks with both an overlay and media background would need the reordered markup, reducing the number of blocks requiring a migration via the deprecation.

renatho replied:
Please, let me know if I'm missing something from your suggestion, but I think we still need it.

The main goal of our change is the `z-index` cleanup, removing especially the z-index from the `.wp-block-cover__inner-container`. But we can do it only with the new markup, otherwise, it would break styles for those that don't re-save their posts.

In other words, the `has-no-zindex-in-inner-blocks` class is more like a versioning class (with a more semantic name than a `vXX`) to check if we have the new markup to apply the new styles.

aaronrobertshaw replied:
>  removing especially the z-index from the .wp-block-cover__inner-container. But we can do it only with the new markup, otherwise, it would break styles for those that don't re-save their posts.

The container could have a style using `:has()` that checks the order of direct children using the `+` sibling selector. If they are in the old order, the style applies the z-index for BC. If the children are in the new order, no z-index would be applied. Another option in our toolkit would be to use `::first-child` when checking the order etc.

As noted in my [other comment](https://github.com/WordPress/gutenberg/pull/66249#discussion_r1812043998), I might not have the bandwidth to give this a try until later this week or early next. Hopefully, this makes sense and gives you a thread to follow though 🧵

renatho replied:
Yes! I could iterate and answered in the other thread. 😉 Thank you for your suggestions!
https://github.com/WordPress/gutenberg/pull/66249#discussion_r1812826334

renatho committed:
Add position relative to keep inner content on the top

renatho committed:
Fix spinner position to be over the overlay

renatho commented on packages/block-library/src/cover/style.scss:
```diff
@@ -96,15 +95,14 @@
-			z-index: auto;
+	&.has-no-zindex-in-inner-blocks {
+		.wp-block-cover__inner-container {
+			position: relative; // It needs a position to be on top of the other elements.
```
Please, look at this property with a grain of salt. I think it won't cause any issue, but if we had an element inside this block with a position absolute, it would be affected.

The next block in the tree with the position relative would be the immediate parent (`.wp-block-cover`). I don't believe we would have an existing think trusting in the position relative to the `.wp-block-cover`, but we never know! =P

aaronrobertshaw replied:
After a quick glance it seems okay. Once the deprecation and styles are iterated on we can test this aspect after they've settled.

renatho commented:
I'm already opening it for review, but it still has some test issues.

I tried to update the fixtures through `npm run fixtures:regenerate`, but I received some errors. If someone could explain me what is happening and how I could fix, it would be very helpful to save some time. =)

aaronrobertshaw commented on packages/block-library/src/cover/editor.scss:
```diff
@@ -118,3 +117,10 @@
 }
+
+// Backward compatibility for v14 deprecation.
+.wp-block-cover:not(.has-no-zindex-in-inner-blocks) {	// Shown while media is being uploaded
+	.components-spinner {
+		z-index: z-index(".wp-block-cover__inner-container");
+	}
+}
```
Do we need to scope this by the proposed CSS class? The old spinner style would result in the same z-index as the inner container, which comes later in the DOM. 

At a quick glance, the positioning of the spinner before the inner container is the same on trunk and this PR. In both cases the spinner is rendered beneath the inner container.

renatho replied:
> Do we need to scope this by the proposed CSS class?

We could keep it as it was. I changed it more as part of the z-index cleanup.
I don't have a strong opinion about this element, so if you have, let me know that I keep this z-index.

> At a quick glance, the positioning of the spinner before the inner container is the same on trunk and this PR. In both cases the spinner is rendered beneath the inner container.

Just notice that if we keep the z-index in the spinner and remove from the inner blocks (where it's important to fix the issue), it would change the original behavior that you described. I didn't test, but I think it would probably be better because I suspect we would always want the spinner in the front.

aaronrobertshaw replied:
I'm not following that last suggestion sorry. Could you clarify it for me?

If you're suggesting that we change the behaviour of the spinner to bring it to the top, I'm not so sure about that. I'd vote that we don't introduce more changes here that aren't required for the bug fix. We should keep trunk's behaviour then make an informed decision on bringing the spinner forward, probably with some design feedback too.

renatho replied:
> Could you clarify it for me?

Thinking more now, that `.wp-block-cover:not(.has-no-zindex-in-inner-blocks)` probably wouldn't be needed because it is in the editor side and it would always be with the latest HTML. But we needed to remove the z-index from the spinner because the inner container doesn't have the z-index anymore, so the spinner would be above it (I suppose it's expected to have the inner container above, like in the `trunk`, so the user can edit the content while the media is being uploaded).

Here are 2 screenshots just for more clarification:

<img width="74" alt="Screenshot 2024-10-23 at 10 39 09" src="https://github.com/user-attachments/assets/9e9afc0e-ae51-4891-b0d4-76055d0a4c09">

<img width="77" alt="Screenshot 2024-10-23 at 10 16 15" src="https://github.com/user-attachments/assets/d40ae026-b50f-45ca-b4eb-c22ded36bea5">

That said, I tweaked it after applying the [new discussed approach](https://github.com/WordPress/gutenberg/pull/66249#discussion_r1812826334): https://github.com/WordPress/gutenberg/pull/66249/commits/e29b238d67fa59af474695e13f3dfb78c52d1a39

We needed to add a [z-index](https://github.com/WordPress/gutenberg/commit/e29b238d67fa59af474695e13f3dfb78c52d1a39#diff-155f0f6f698abb06228620a990f9efb18dceb8f1abf883a61c4c47ecbcf1cdbfR42) in the inner container during the transient state (uploading) because of the [transient overlay](https://github.com/WordPress/gutenberg/commit/e29b238d67fa59af474695e13f3dfb78c52d1a39#diff-155f0f6f698abb06228620a990f9efb18dceb8f1abf883a61c4c47ecbcf1cdbfR38). It means that we would have the original issue (the cover appearing over the navigation overlay) in this moment, but I suspect it's fine since it's just this quick moment and it's only in the editor side. WDYT? If you consider it bad, we could try to explore other approaches to change the transient overlay, but I don't think it's worth it.

aaronrobertshaw replied:
I think that could be something we can iterate on in a follow-up.

renatho replied:
Makes sense! I suspect it would be a very low priority (I doubt a user would reproduce it 🤭), so I can open an issue after we merge this. 😉

aaronrobertshaw commented on packages/block-library/src/cover/style.scss:
```diff
@@ -302,3 +292,32 @@ section.wp-block-cover-image > h2,
 }
+
+// Backward compatibility for v14 deprecation.
+.wp-block-cover:not(.has-no-zindex-in-inner-blocks) {
```
As noted in the [above thread](https://github.com/WordPress/gutenberg/pull/66249/files#r1806691866), it could be possible to avoiding needing `.has-no-zindex-in-inner-blocks` completely.

aaronrobertshaw commented on packages/block-library/src/cover/style.scss:
```diff
@@ -302,3 +292,32 @@ section.wp-block-cover-image > h2,
+	.wp-block-cover__inner-container {
+		z-index: z-index(".wp-block-cover__inner-container");
+
+		// Reset the z-index to auto when the body has a modal open. So when the
+		// modal is inside the cover, it doesn't create a stacking context.
+		@at-root .has-modal-open & {
+			z-index: auto;
+		}
```
If https://github.com/WordPress/gutenberg/pull/66093 isn't slated for release until 19.6, could we also avoid these additional styles? I don't think we need to provide BC for a change that hasn't been released yet.

renatho replied:
I think it's good to have this anyway, so it's automatically fixed in all the sites just by updating the Gutenberg. Make sense?

aaronrobertshaw replied:
Maybe my memory is failing me but the original issue was in the editor only correct? The interactivity API was used on the frontend and that was all displaying ok.

In the editor, with a deprecation in place, the blocks would be migrated and thus fixed without needing the superfluous styles. So the problem would still be automatically fixed by updating Gutenberg only without the unnecessary styles and the block modifying document level classes.

renatho replied:
The main issue was on the frontend (we had a few reports about that from users). And as a lower impact, it was also happening in the editor.

Some reports from users can be found in:

- https://github.com/Automattic/wp-calypso/issues/75439
- https://github.com/Automattic/wp-calypso/issues/79317

aaronrobertshaw replied:
Thanks for setting me straight on that. 

So, we can leave this style there but the [`has-modal-open` class being added](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/edit/index.js?plain=1#L87) to the document wrapper should be removed. The deprecation should sort out the migration and display of the block in the editor so we no longer need the change that was introduced in https://github.com/WordPress/gutenberg/pull/66093.

Avoiding that was the main driver for this whole approach in my view.

renatho replied:
I thought the same, but I decided to keep given this reason:

> As part of this PR, I also thought of removing [this effect that adds the has-modal-open in the editor](https://github.com/WordPress/gutenberg/pull/66093/files#diff-884dae78f1676d454c049107e9b05b2403eb042457f77f836bd170d104396b8bR89-R97) because that wouldn't be needed for this purpose anymore. But thinking more, I kept it to keep the consistency there between the frontend and the editor.

If you reconsider, and think it's still worth removing it, I can remove it. But I think it would be better to keep this for consistency (frontend - editor) for the future.

> Avoiding that was the main driver for this whole approach in my view.

I think the benefit is that we only use it for BC and not as our main solution anymore.

aaronrobertshaw replied:
> If you reconsider, and think it's still worth removing it, I can remove it. But I think it would be better to keep this for consistency (frontend - editor) for the future.

To me, a block mutating the classes on the overall document is very odd in the editor. It also attracted similar reactions from other contributors on that original PR.

I appreciate the consistency argument and I would normally lean that way myself. In this scenario though, without a genuine use case or purpose, it would be best not to keep that. It could also encourage improper coupling of blocks, other components etc. to that class. 

If the situation changes in the future we could always revisit bringing in that consistency. That might even be via the interactivity API itself and be truly consistent with the frontend then. This approach would also then allow a dedicated discussion around that behaviour without a rush to a decision.

TL;DR - Let's remove the addition of the `has-modal-open` class for now.

renatho replied:
Gotcha! Reverted here: https://github.com/WordPress/gutenberg/pull/66249/commits/7226f4bd33ed00ca3c2effa254c8c31e7aedd2a7

I'll probably continue tomorrow. And thanks for the commits you added here. 😉

andrewserong replied:
Looks like the styles are still to be removed?

renatho replied:
@andrewserong , I think you are referring to the `@at-root .has-modal-open &`, right?

If yes, I think it's good to keep it for backward compatibility. With this, it gets fixed in the frontend for the users without needing to re-save the post.

aaronrobertshaw commented on packages/block-library/src/cover/deprecated.js:
```diff
@@ -244,6 +244,157 @@ const v12BlockSupports = {
 };
 
+// Deprecation for blocks that have z-index.
+const v14 = {
```
This deprecation will need a fixture to cover it. Here's an [example](https://github.com/WordPress/gutenberg/blob/trunk/test/integration/fixtures/blocks/core__cover__deprecated-2.html?plain=1#L1) for an earlier deprecation.

renatho replied:
Thank you for the suggestion and the example! ♥️
I'll try to fix it today!

renatho replied:
I followed [this documentation](https://github.com/WordPress/gutenberg/blob/trunk/test/integration/fixtures/blocks/README.md), and based on your example, I added the new fixture: https://github.com/WordPress/gutenberg/pull/66249/commits/9ba72c00bdebe83eaf5e844e2b9fe178096d3849 (created the HTML and let the command generate the others)

I'm still struggling a bit to fix the others. I let you know here when it's done.

renatho replied:
I fixed most of the issues (https://github.com/WordPress/gutenberg/pull/66249/commits/0ae62731398295ede211eb9d199b4853ddcfea0c), but in the fixture tests it's trying to add the `'has-no-zindex-in-inner-blocks'` as a `className` attribute.

Not sure if it's a bug with the fixture tests, or if adding a class to the wrapper element wouldn't be recommended. 🤔

I thought of not using the sibling class, as you suggested to fix this, but it would mean a few new classes instead of just one. Another approach would be adding an additional wrapper with a class, it would break a few styles (we have some using `>` that we would need to refactor to do it, so I think it would be risky).

I'm out of ideas now. I need to think more.

And I noticed that I also need to update some mobile snapshots.

aaronrobertshaw replied:
Thanks for the updates, I'll review those shortly. 

Regarding the sibling selectors though, why would we need additional classes? 

We should be able to rely on the order of the elements to have the correct sibling selector apply because the old and new versions moved the overlay element. When there aren't multiple elements there, we wouldn't need to juggle their z-index. 

I'm a little short on bandwidth at the moment to code something up but I could take a run at it later this week perhaps if you don't beat me to it.

It could be worth pursuing this, even if it is ruled out as an option, as it would limit the number of blocks needing migration and avoid the slightly awkward class on the wrapper.

renatho replied:
Hey! 👋

Thank you for your suggestions! I followed your suggestion and applied it with this commit: https://github.com/WordPress/gutenberg/pull/66249/commits/0e25fed09bdf79a48f0f32a678c36da5a19d056e.

aaronrobertshaw reviewed (commented):
Thank you for following up on https://github.com/WordPress/gutenberg/pull/66093 @renatho 👍 

I've taken this for a test drive. It's coming together well but I did run into a few issues.

✅ Cover z-index behaviour works as expected post migration
✅ Deprecated blocks appear to migrate in the editor
✅ Deprecated blocks still render appropriately on the frontend
❌ After saving the post and reloading, the deprecated blocks don't appear to have been migrated i.e. they retrigger the deprecation. Making an unrelated edit and resaving the post seems to apply the updates the second time around.


https://github.com/user-attachments/assets/12260ea5-702a-4263-9dad-e6c52534ec24



In addition to the migration issue flagged above, I've left a few other thoughts as inline comments where applicable.

These include:

1. We should be able to avoid introducing the new CSS class to the markup by using CSS sibling selectors for both the latest and BC styles.
   - This would also mean that less blocks would need to be migrated, perhaps mitigating a tiny bit of the performance hit.
   - i.e. only blocks with both the color overlay and an image or video would have their markup changed requiring a migration.
2. As the original fix is slated for release in 19.6, we could avoid the need to include styles for `.has-modal-open` for BC reasons. 
    - I understand the consistency argument but simply including it for no functional reason and setting a precedent for a block in the editor to mess with the document root classes feels like a stretch. 
    - Happy to hear others' thoughts on that one. My vote is remove it, we can always add it when there is a pressing need.
3. We need a new fixture to cover this latest deprecation.
4. All the non-deprecation-related fixtures for the cover block need to be updated to the latest markup structure. 
    - This is why the fixture regeneration is throwing the error around an unexpected deprecation being run.
6. After the above fixture updates, regenerating the fixtures should succeed and the result should be something like past deprecation's serialized HTML being updated to the correct structure.

I suspect that in the process of ironing out the deprecation, the bug I noticed while testing will be uncovered and resolved 🤞 

@renatho let me know if anything around my recommendations for the deprecation tweaks isn't clear.

renatho committed:
Add new fixture

renatho committed:
Update cover fixtures

renatho committed:
Update deprecation to match last update from trunk

renatho commented:
> ❌ After saving the post and reloading, the deprecated blocks don't appear to have been migrated i.e. they retrigger the deprecation. Making an unrelated edit and resaving the post seems to apply the updates the second time around.

@aaronrobertshaw, could this be related to the way it works when using the code editor mode? I could reproduce the same, but it seems it doesn't change the blocks when we are in the code editor mode. If we are in the visual mode and we edit anything and save, it changes properly, even if it's on the first try.

renatho committed:
Add class to the editor

aaronrobertshaw commented:
> could this be related to the way it works when using the code editor mode?

Good point. I'll give it a bit more of a test. We'll also need to ensure that it is migrated correctly if patterns containing a deprecated version of the cover block are inserted.

aaronrobertshaw reviewed (commented):
Thanks for updating the deprecation fixtures @renatho 💪 

That's a lot of files! 😅 

Unfortunately, it looks like a large number shouldn't have needed changes. 

> 4. All the non-deprecation-related fixtures for the cover block need to be updated to the latest markup structure

I'm sorry I wasn't clearer with this comment in my last review.

All the source fixtures for deprecations e.g. `*_deprecated-#.html` should remain as before. These fixtures are supposed to cover the deprecated versions of the block, as the were when they were deprecated. 

Only the source fixtures for the current version of the block needed to be updated to the latest markup structure. These are all the `.html` files that don't have `__deprecated-#` in their name.

> 5. After the above fixture updates, regenerating the fixtures should succeed and the result should be something like past deprecation's serialized HTML being updated to the correct structure.

With the source fixtures being correct, regenerating the fixtures would mean that the `_deprecated-#.serialized.html` files all get updated showing they successfully migrate to the latest version i.e. with the wrapper's class (if it's kept) and the reordered inner elements as appropriate.

--- 

It might not seem like it but I don't think this PR is that far off. The deprecation itself seems ok, we just need to sort out the fixtures and the CSS approach.

I'd recommend to iterate on the CSS first as the possible removal of the need for a new CSS class would reduce the number of deprecation fixtures requiring change. Doing it the other way around would likely create unnecessary rework.

Thanks again for all your hard work and patience here @renatho 🙇

renatho committed:
Remove class in favor of a new approach checking the children

renatho committed:
Use styles only on frontend for backward compatibility

renatho committed:
Fix loading state

renatho commented:
Thank you for all the quick and priceless reviews, @aaronrobertshaw! And thanks for your patience! =)

> All the source fixtures for deprecations e.g. *_deprecated-#.html should remain as before. These fixtures are supposed to cover the deprecated versions of the block, as the were when they were deprecated.
> 
> Only the source fixtures for the current version of the block needed to be updated to the latest markup structure. These are all the .html files that don't have __deprecated-# in their name.

Thanks! That makes more sense now! I'll wait to fix them after we have the rest done. :)

aaronrobertshaw committed:
Revert "Update cover fixtures"

This reverts commit 0ae62731398295ede211eb9d199b4853ddcfea0c.

aaronrobertshaw committed:
Fix deprecations and fixtures

aaronrobertshaw commented:
Nice work implementing the alternate CSS approach @renatho, it's working well for me in my local testing ✨ 

| Frontend | Editor |
|---|---|
| <video src="https://github.com/user-attachments/assets/2c3e2ebd-e9df-4ef0-973d-44d4123441e8" /> | <video src="https://github.com/user-attachments/assets/4eb729b1-4916-4a68-b191-87e5e647edd3" /> |



Given the complexities of deprecations, it might be easier to outline some of the problems this PR had with them against the backdrop of some commits that illustrate the proper updates. So in that spirit, I hope you don't mind, I've pushed two commits.

The [first](https://github.com/WordPress/gutenberg/pull/66249/commits/71b3ae88a76099bddd101bab7972e52d1b5281f0) simply reverts all the past updates as there were about 34 files that didn't need to be touched, along with others that were updated incorrectly. The [second commit](https://github.com/WordPress/gutenberg/pull/66249/commits/03ac4bff709a524c55b8f5914560adda7d0be602) is the one that might be of interest and that's the one I'd suggest to read along with my findings below. Hopefully, the combination will make sense.

- I started out by fixing up the non-deprecation fixtures or "default" fixtures, whatever you want to call them
    - Included updating the `.html` files to the latest version's markup
    - After re-running the regenerate fixtures command, the `parsed.json` and `serialized.html` files were updated appropriately. The `.json` file isn't updated as the attributes haven't changed.
    - Default fixtures updated were:
        - `core__cover.html`
        - `core__cover__alt-fixed-background.html`
        - `core__cover__duotone-fixed-background.html`
        - `core__cover__gradient-image.html`
        - `core__cover__gradient-video.html`
        - `core__cover__video-overlay.html` (Seems this one had incorrect attributes in `.json` file though)
        - `core__cover__video.html`
- Then, I updated the deprecation fixtures that required it. These were: 1-8, 10, and the new v14 fixture.
- The new deprecation fixture for v14 was incorrect and needed the source to be updated
    - The fixture previously had the latest version's markup. Not the deprecated markup that the deprecation fixture was supposed to represent
    - It also appeared that the `isUserOverlayColor` attribute was missing from the deprecation attributes in `deprecated.js`
- There's also still one discrepancy with the deprecation attributes for the `alt` attribute that I haven't had the time to resolve
    - The `alt` attribute in the deprecations does not match the definition within the block.json file.
    - It appears that at some stage the attribute was migrated away from using the markup as its source.
    - To fix this we need to pinpoint which version it actually changed in and correct the deprecation attributes
    - The affect of this issue is those blocks that used the markup as the source previously might be losing that data when they get migrated
- The new v14 deprecation also was using the outdated `v12BlockSupports` definition
    - This means that data around shadows, aspectRatio, and interactivity API's clientNavigation would be lost when migrating.
    
With the commits I've pushed, all the deprecations are passing. All that is left is:
1. Fix the `alt` attribute in deprecations
2. Get some fresh eyes to compare the current attributes/supports/save definitions with what's in the deprecation

To that last point, @andrewserong if you could find some time to lend your eagle eye to this endeavour, that would be appreciated 🙏 

Also, can this PR be backported to 6.7 as a bug fix or would it only be considered an enhancement because it replaces and improves the shorter-term bug fix that already landed? If it doesn't get backported what would it mean for BC? 

My brain is a little melted after untangling the deprecations here so there might be easy answers I'm missing.

@renatho if there's anything that doesn't make sense on the commits I pushed, let me know. I'm happy to explain anything at all 👍

andrewserong commented:
> if you could find some time to lend your eagle eye to this endeavour, that would be appreciated

Thanks for the work here both of you! I mightn't get a chance to look this over properly today, but I've got it on my list for tomorrow 👍

renatho committed:
Revert "Add the has-modal-open to the editor reproducing the same behavior of the frontend"

This reverts commit 64c14d3d4ff081a07bbb8583812650992b9138e0.

andrewserong commented on packages/block-library/src/cover/style.scss:
```diff
@@ -95,17 +94,11 @@
 	}
 
 	.wp-block-cover__inner-container {
+		position: relative; // Needed after the v14 deprecation..
```
Tiny nit: can we explain why it's needed? The deprecations can be hard to understand, so if I read this comment I'd need to go looking to figure it out. Alternately, we could remove the comment if an explanation isn't necessary.

renatho replied:
Improved as part of this commit: https://github.com/WordPress/gutenberg/commit/f0640dbded1b7d0aa6baf708a7079bdabfb86588

Thanks for the suggestion!

andrewserong commented on packages/block-library/src/cover/style.scss:
```diff
@@ -302,3 +287,41 @@ section.wp-block-cover-image > h2,
+	`.wp-block-cover__background` , it means that it's using the new HTML
+	structure or that it's using the color background or gradient (even in
+	the deprecated version). In the last option we also don't need the z-index
+	because the elements are in the expected order. */
```
Should we use `//` comments so that the this comment doesn't get included in the built CSS?

renatho replied:
The problem is that it's a comment in the middle of the selector.
But I moved it and although complex, I hope that's clear enough. I also added another comment in the `save.js` just to make sure we don't break anything accidentally.

https://github.com/WordPress/gutenberg/pull/66249/commits/f0640dbded1b7d0aa6baf708a7079bdabfb86588

andrewserong reviewed (commented):
This is testing great for me so far, nice work both of you on honing in on the deprecation!

✅ Cover block continues to work as expected with a nested Navigation block, and the modal not erroneously displaying other Cover blocks over the top of it
✅ Double-checked that Group block variations still work correctly (that [the previous regression](https://github.com/WordPress/gutenberg/issues/66067) wasn't introducde)
✅ Old Cover blocks render correctly on the site frontend as far as I can tell
✅ Old Cover blocks migrate correctly in the editor

Overall I like the approach here of using a different DOM order and removing the `z-index`es. In general where possible I like it when we can avoid adding deprecations as we've already got a lot for this block and it can be hard to reason about. But in this case, I think re-ordering and removing the `z-index` rules makes it worth it 👍

I see the mobile tests are still failing and will need an update. I left a couple of tiny comments. What else is there left to do before this is ready for a final review? (I see Aaron listed some above, but I thought I'd double check) Happy to keep testing next week!

andrewserong commented:
> Get some fresh eyes to compare the current attributes/supports/save definitions with what's in the deprecation

I forgot to add in my previous comment — I looked and compared these and the deprecation's `save` function matches what was previously in `save.js` and the added attributes and supports appears to cover everything since the last deprecation was made, as far as I can tell from a little git history 👍

renatho committed:
Improve documentation comments in the code

renatho committed:
Fix native tests

renatho committed:
Update snapshots

renatho commented:
I fixed the React Native tests:

- https://github.com/WordPress/gutenberg/pull/66249/commits/219b65fec212508c349d13fdc1d103225ffb3f89
- https://github.com/WordPress/gutenberg/pull/66249/commits/99aa8f90b29c3b835117d657d96392a61159647a

It seems all the tests are passing now, but when I run the `npm run fixtures:regenerate` it reorders some properties (`useFeaturedImage`, `isUserOverlayColor`, `tagName`) in `core__cover__deprecated-10.json`, `core__cover__deprecated-14.json` and `core__cover__video-overlay.json`. I wonder if we should commit this. Does this also happen for you @aaronrobertshaw? After updating, if I run it again, it just passes with no more changes.

> There's also still one discrepancy with the deprecation attributes for the alt attribute that I haven't had the time to resolve

I see it. I investigated it, and solved it with this commit: https://github.com/WordPress/gutenberg/pull/66249/commits/330fabe67c9e525f9f4a5317a16ed019d47f60c6

Basically, [this PR](https://github.com/WordPress/gutenberg/commit/af34a42bd3fa1045fcd4b9d7542f451aa27ce620#diff-384718dc529cfe56aecf02f99fa654c6937c8d8d35cbc0827863ecaef6eb73ae) changed it but didn't update the deprecated block attributes.

aaronrobertshaw commented:
> Does this also happen for you @aaronrobertshaw?

From memory, this is due to the way the deprecation attribute objects are created i.e. spreading the previous version's attributes then adding anything, resulting in a different order.

There are a few options:
1. Make the deprecation attributes an actual 1:1 with that version's (including order)
2. Don't bother with updating the parsed fixture files when they are just a reorder of attributes
3. Commit them and accept the extra noise

andrewserong commented:
> There are a few options:

Thanks for laying that out! I'd go for either 1 or 3 so that the next person who runs the command to update fixtures isn't confused as to why there are changes to those files. If it improves readability to not worry about dealing with option 1, then option 3 could be the simplest path forward?

aaronrobertshaw reviewed (commented):
Thanks for the continued iteration here @renatho and the additional reviews @andrewserong. 

> I see it. I investigated it, and solved it with this commit: https://github.com/WordPress/gutenberg/commit/330fabe67c9e525f9f4a5317a16ed019d47f60c6
> 
> Basically, [this PR](https://github.com/WordPress/gutenberg/commit/af34a42bd3fa1045fcd4b9d7542f451aa27ce620#diff-384718dc529cfe56aecf02f99fa654c6937c8d8d35cbc0827863ecaef6eb73ae) changed it but didn't update the deprecated block attributes.

Nice work isolating that. I did mean to link that PR as a thread for you to follow, sorry.

I think we still have an issue though with the deprecation.

Note in the PR you linked that it did two things:
1. Altered the block's `alt` attribute
2. Created a deprecation: `v13`

That deprecation (`v13`) needed to migrate the previous version of the block. The change to the `alt` attribute definition was for (at that time) the current block version not the previous one.

So in other words, that PR "didn't miss anything" in relation to the attributes.

The error was introduced in this PR when it didn't include the updated `alt` attribute definition to migrate the previous version to the latest. With the latest commits, we've swapped one problem for another. 

The `v13` deprecation now includes the updated `alt` attribute definition when it shouldn't. The result would be blocks that should have been migrated by that deprecation will lose their custom alt text.

All this nuance, is a big part of why block deprecations are especially tricky to work with. The good news is, you're nearly there and the next deprecation is going to seem simple after wrestling the Cover block into submission 💪 .



---

Regarding the fixtures' parsed attributes being reordered, it [appears](https://github.com/WordPress/gutenberg/pull/66249#issuecomment-2440264361) that you should be fine to commit those. The order of attributes in these files already doesn't match the actual real block's when parsed but I don't think that should have any consequence.

renatho committed:
Fix order of the properties on the generated fixtures

renatho committed:
Add alt attribute as part of the attributes changed in the previous version

renatho commented:
> That deprecation (v13) needed to migrate the previous version of the block. The change to the alt attribute definition was for (at that time) the current block version not the previous one.

Hm! Good catch!

> All this nuance, is a big part of why block deprecations are especially tricky to work with.

Yeah! It definitely seems tricky. Especially because we need to investigate the history of the block to deprecate it correctly. Thank you for all the learning here! Next time when I need to do it, it will be easier with a little more knowledge!

I removed the previous commit: https://github.com/WordPress/gutenberg/pull/66249/commits/330fabe67c9e525f9f4a5317a16ed019d47f60c6
And added this one: https://github.com/WordPress/gutenberg/pull/66249/commits/92cc116687ba53e764d0d436ce87dc84b752be4b

So the new version of the `alt` is changed only now, when the previous `alt` should already be the new one. I think it looks correct now, right? So this is correct for the next versions.

> Regarding the fixtures' parsed attributes being reordered, it https://github.com/WordPress/gutenberg/pull/66249#issuecomment-2440264361 that you should be fine to commit those. The order of attributes in these files already doesn't match the actual real block's when parsed but I don't think that should have any consequence.

Commit here: https://github.com/WordPress/gutenberg/pull/66249/commits/ce1399bd8e467da3dcff8ed0e655a898bd8ef5fa

aaronrobertshaw commented on packages/block-library/src/cover/test/__snapshots__/edit.native.js.snap:
```diff
@@ -33,72 +33,72 @@ exports[`color settings toggles between solid colors and gradients 1`] = `
 exports[`minimum height settings changes the height value between units 1`] = `
-"<!-- wp:cover {"url":"https://cldup.com/cXyG__fTLN.jpg","id":10710,"dimRatio":50,"overlayColor":"foreground","isUserOverlayColor":true,"minHeight":50,"minHeightUnit":"px","isDark":false} -->
-<div class="wp-block-cover is-light" style="min-height:50px"><span aria-hidden="true" class="wp-block-cover__background has-foreground-background-color has-background-dim"></span><img class="wp-block-cover__image-background wp-image-10710" alt="" src="https://cldup.com/cXyG__fTLN.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeh ...
+"<!-- wp:cover {"url":"https://cldup.com/cXyG__fTLN.jpg","id":10710,"dimRatio":50,"overlayColor":"foreground","minHeight":50,"minHeightUnit":"px","isDark":false} -->
```
How'd you generate these updates? 

I'm wondering why the `isUserOverlayColor` attribute was removed.

renatho replied:
I used `npm run test:native:update` to update the snapshots.

aaronrobertshaw commented on packages/block-library/src/cover/test/edit.native.js:
```diff
@@ -54,12 +54,12 @@ const COVER_BLOCK_SOLID_COLOR_HTML = `<!-- wp:cover {"overlayColor":"cyan-bluish
 <!-- /wp:cover -->`;
 const COVER_BLOCK_IMAGE_HTML = `<!-- wp:cover {"url":"https://cldup.com/cXyG__fTLN.jpg","id":10710,"dimRatio":50,"overlayColor":"foreground","isDark":false} -->
-<div class="wp-block-cover is-light"><span aria-hidden="true" class="wp-block-cover__background has-foreground-background-color has-background-dim"></span><img class="wp-block-cover__image-background wp-image-10710" alt="" src="https://cldup.com/cXyG__fTLN.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","f ...
+<div class="wp-block-cover is-light"><img class="wp-block-cover__image-background wp-image-10710" alt="" src="https://cldup.com/cXyG__fTLN.jpg" data-object-fit="cover"/><span aria-hidden="true" class="wp-block-cover__background has-foreground-background-color has-background-dim"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","f ...
```
I think these need updating further.  This change brings the HTML markup only up-to-date. It's missing the `isUserOverlayColor` true attribute which might be causing the removal of that attribute in the snaps that have also been updated.

renatho replied:
Thank you for your careful reviews and your patience, @aaronrobertshaw!

I updated these commits to fix this: https://github.com/WordPress/gutenberg/pull/66249/commits/eaac4cd6505712044cee563b188c78e028840da7 https://github.com/WordPress/gutenberg/pull/66249/commits/9ae33452a31490868a47a6f204adfc257cd5809e

Honestly, I didn't investigate deeper the reasons to understand better, but I suspect it's because we added this attribute to the deprecation now (`v14BlockAttributes`). If you are not confident that it's really needed and we still need some investigation, let me know and I can investigate this more another time. 😉

aaronrobertshaw reviewed (commented):
At first glance I thought we'd made it to the finish line. Unfortunately, it looks like there's one small issue around the updates to the native tests and their related snaps.

I've left inline comments for those below. Once they are taken care of, we should be ready to merge 🤞 

---

> Especially because we need to investigate the history of the block to deprecate it correctly. 

This isn't the case in my experience.

The majority of the issues we hit here were due to not working from the current block's attribute and support definitions etc. This [commit](https://github.com/WordPress/gutenberg/pull/66249/commits/1c2837872a5b6c62f62466d6abc77bb5835d8bc2) shows the new deprecation worked from the last deprecation not the current block. 

The other complicating factor was it not being intuitive as to how to update the deprecation fixtures correctly.

So to me, the "historical dig" was to just confirm where and when the deprecation went off the rails.

aaronrobertshaw reviewed (approved):
Nice work @renatho, thanks for the determination to stick with this 🙇 

This PR is looking good and testing well for me.

✅ Unit tests are passing
✅ Fixtures and deprecations look good
✅ Cover block is rendering as expected both before and after migration
✅ Didn't spot any regressions around the original issue with nav block overlays

Given the bumps in the road getting here, I think it would be prudent to get a fresh second set of eyes on this before merging. (cc @andrewserong when you have the time)

From my end though, this gets a ✅ as good to go.

andrewserong commented:
Thanks for the ping @aaronrobertshaw and for all the detailed reviews here. I'll give it a review in the next few hours!

andrewserong reviewed (approved):
This is testing nicely for me, too, I did a couple of other random smoke tests of things related to z-indexes and haven't run into any issues!

✅ Cover blocks with nested Navigation blocks are working as expected before and after applying this PR
✅ Navigation block overlays appear to work correctly in the post and site editors and the site frontend
✅ Deprecations appear to apply correctly
✅ Other configurations (e.g. video backgrounds, rounded borders) work correctly
✅ Sticky Group blocks within Cover blocks work just as on `trunk`
✅ Sticky Group blocks outside of Cover blocks work just as on `trunk`
✅ Deprecations and updated fixtures and tests look good as far as I can tell

Nice work persevering with this one @renatho, this LGTM! 🚀

andrewserong commented:
Looks like we're all happy with this, so I'll merge it in. Thanks again for all the work here, both of you!

aaronrobertshaw commented:
Thanks for getting this one merged @andrewserong 🚀
